hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOffloadableBackupOperation.java

```diff
@@ -0,0 +1,211 @@
+import java.util.Map;
+
+import static com.hazelcast.config.InMemoryFormat.OBJECT;
+import static com.hazelcast.core.EntryEventType.ADDED;
+import static com.hazelcast.core.EntryEventType.REMOVED;
+import static com.hazelcast.core.EntryEventType.UPDATED;
+import static com.hazelcast.spi.ExecutionService.ASYNC_EXECUTOR;
+
+/**
+ * Backup processing for the EntryOffloadableOperation
+ *
+ * See the javadoc on {@link EntryOffloadableOperation}
+ */
+public class EntryOffloadableBackupOperation extends LockAwareOperation implements BackupOperation, MutatingOperation {
+
+    private EntryBackupProcessor entryProcessor;
+
+    public EntryOffloadableBackupOperation() {
+    }
+
+    public EntryOffloadableBackupOperation(String name, Data dataKey, EntryBackupProcessor entryProcessor) {
+        super(name, dataKey);
+        this.entryProcessor = entryProcessor;
+    }
+
+    @Override
+    public void innerBeforeRun() throws Exception {
+        super.innerBeforeRun();
+
+        if (entryProcessor instanceof HazelcastInstanceAware) {
+            HazelcastInstance hazelcastInstance = getNodeEngine().getHazelcastInstance();
+            ((HazelcastInstanceAware) entryProcessor).setHazelcastInstance(hazelcastInstance);
+        }
+    }
+
+    @Override
+    public void run() {
+        final long now = getNow();
+        SerializationService serializationService = getNodeEngine().getSerializationService();
+        Object oldValue = recordStore.get(dataKey, true);
+
+        if (!(entryProcessor instanceof Offloadable)) {
+            throw new HazelcastException("BackupEntryProcessor is expected to be Offloadable.");
+        }
+        if (entryProcessor instanceof ReadOnly) {
+            throw new HazelcastException("BackupEntryProcessor cannot be ReadOnly.");
+        }
+
+        boolean shouldCloneForOffloading = OBJECT.equals(mapContainer.getMapConfig().getInMemoryFormat());
+        oldValue = shouldCloneForOffloading ? serializationService.toData(oldValue) : oldValue;
+        runOnOffloadedThread(oldValue, ((Offloadable) entryProcessor).getExecutorName(), now);
+
+    }
+
+    @SuppressWarnings("unchecked")
+    private void runOnOffloadedThread(final Object oldValue, String executorName, final long now) {
+        final OperationServiceImpl ops = (OperationServiceImpl) getNodeEngine().getOperationService();
+
+        final Data finalDataKey = dataKey;
+        final String finalCaller = getCallerUuid();
+        final long finalThreadId = threadId;
+        final long finalCallId = getCallId();
+
+        ops.onStartAsyncOperation(this);
+        executorName = executorName.equals(Offloadable.DEFAULT_ASYNC_EXECUTOR) ? ASYNC_EXECUTOR : executorName;
+
+        final ObjectNamespace namespace = new DefaultObjectNamespace(MapService.SERVICE_NAME, name);
+
+        // TODO hope this will lock the backup on a Backup Operation?
+        recordStore.lock(finalDataKey, finalCaller, finalThreadId, finalCallId, Long.MAX_VALUE);
```

Comment:
Partition can get migrated during offloaded execution and unlock invocation after the processing can fail (either unlock fails or timeouts and you can't learn the result). 

Even record may be unlocked on primary but remain locked on backup. And when backup is promoted, it will remain locked forever.

So, I think it's better to acquire the lock with a lease.